src/main/scala/com/fulcrumgenomics/umi/DuplexConsensusCaller.scala

```diff
@@ -140,84 +151,99 @@ class DuplexConsensusCaller(override val readNamePrefix: String,
-        val duplexR1 = duplexConsensus(abR1Consensus, baR2Consensus, duplexR1Sources)
-        val duplexR2 = duplexConsensus(abR2Consensus, baR1Consensus, duplexR2Sources)
-
-        // Convert to SamRecords and return
-        (duplexR1, duplexR2) match {
-          case (Some(r1), Some(r2)) =>
-            Seq(
-              createSamRecord(r1, FirstOfPair, toUmiBases(duplexR1Sources)),
-              createSamRecord(r2, SecondOfPair, toUmiBases(duplexR2Sources))
-            )
-          case _                    =>
-            // NB: some reads may have been rejected already in filterToMostCommonAlignment, so just
-            //     reject those records that survived the initial filtering.
-            val remainingRecs = filteredXs ++ filteredYs
-            rejectRecords(remainingRecs.flatMap(_.sam), FilterSsConsensus)
-            Nil
-        }
-      case _ =>
-        unreachable("SamRecords supplied with more than two distinct MI values.")
+  /** Returns true if there enough reads according to the minReads option. */
+  private def hasMinimumNumberOfReads(groups: Seq[Seq[SamRecord]]): Boolean = {
+    // Get the number of reads per strand, in decreasing (more stringent) order.
+    val (numXY: Int, numYX: Int) = {
+      require(groups.length <= 2, "SamRecords supplied with more than two distinct MI values.")
+      val numAb = groups.headOption.map(_.count(r => r.paired && r.firstOfPair)).getOrElse(0)
+      val numBa = groups.drop(1).headOption.map(_.count(r => r.paired && r.firstOfPair)).getOrElse(0)
+      if (numAb >= numBa) (numAb, numBa) else (numBa, numAb)
+    }
+    this.minTotalReads <= numXY + numYX && this.minXYReads <= numXY && this.minYXReads <= numYX
+  }
+
+  /** Attempts to call a duplex conesnsu reads from the two sets of reads, one for each strand. */
+  private def callDuplexConsensusRead(ab: Seq[SamRecord], ba: Seq[SamRecord]): Seq[SamRecord] = {
+    // Fragments have no place in duplex land (and are filtered out previously anyway)!
+    val (_, abR1s, abR2s) = subGroupRecords(ab)
+    val (_, baR1s, baR2s) = subGroupRecords(ba)
+
+    // Get all the alignments to one end of the source molecule
+    val singleStrand1 = abR1s ++ baR2s
+    val singleStrand2 = abR2s ++ baR1s
+
+    // The orientation of AB and BA reads should be:
+    // AB R1: +  AB R2: -
+    // BA R1: -  BA R2: +
+    // or vice versa (AB-R1:-, AB-R2:+, AB-R1:-, AB-R2: +
+    // Therefore, AB-R1s and BA-R2s should be on the same strand, and the same for AB-R2s and BA-R1s
+    // Check for this explicitly here.
+    if (singleStrand1.nonEmpty) {
+      val ss1Flag = singleStrand1.head.negativeStrand
+      if (singleStrand1.exists(_.negativeStrand != ss1Flag)) {
+        val ss1MI   = singleStrand1.head.apply[String](ConsensusTags.MolecularId)
+        rejectRecords(ab ++ ba, FilterCollision)
+        logger.debug(s"Not all AB-R1s and BA-R2s were on the same strand for molecule with id: $ss1MI")
+        return Nil
```

Comment:
`return` is bad: https://tpolecat.github.io/2014/05/09/return.html

Reply:
I think probably you could extract a small function `areAllSameStrand(rs: Seq[SamRecord])` and then restructure this something like:

```scala
(areAllSameStrand(singleStrand1), areAllSameStrand(singleStrand2)) match {
  case (true, true) => 
    // do all the things
  case (result1, result2) => 
    rejectRecords(ab ++ ba, FilterCollision)
    if (!result1) ....
    if (!result2) ....
    Nil
}
```

Reply:
This is how it was previously written.

Reply:
Fixed